Sources/Sentry/SentryMetricProfiler.mm

```diff
@@ -0,0 +1,220 @@
+NSString *const kSentryMetricProfilerSerializationUnitMemoryPressureEnum = @"memory-pressure-enum";
+NSString *const kSentryMetricProfilerSerializationUnitThermalStateEnum = @"thermal-state-enum";
+NSString *const kSentryMetricProfilerSerializationUnitPercentage = @"percent";
+
+namespace {
+NSDictionary<NSString *, id> *
+serializedValues(NSArray<NSDictionary<NSString *, NSNumber *> *> *values, NSString *unit)
+{
+    return @ { @"unit" : unit, @"values" : values };
+}
+} // namespace
+
+@implementation SentryMetricProfiler {
+    NSTimer *_timer;
+
+    SentryNSProcessInfoWrapper *_processInfoWrapper;
+    SentrySystemWrapper *_systemWrapper;
+
+    /// arrays of readings keyed on NSNumbers representing the core number for the set of readings
+    NSMutableDictionary<NSNumber *, NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *>
+        *_cpuUsage;
+
+    NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *_memoryFootprint;
+    NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *_thermalState;
+    NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *_powerLevelState;
+    NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *_memoryPressureState;
+    uint64_t _profileStartTime;
+}
+
+- (instancetype)initWithProfileStartTime:(uint64_t)profileStartTime
+                      processInfoWrapper:(SentryNSProcessInfoWrapper *)processInfoWrapper
+                           systemWrapper:(SentrySystemWrapper *)systemWrapper
+{
+    if (self = [super init]) {
+        _cpuUsage = [NSMutableDictionary<NSNumber *,
+            NSMutableArray<NSDictionary<NSString *, NSNumber *> *> *>
+            dictionary];
+        const auto processorCount = NSProcessInfo.processInfo.processorCount;
+        for (NSUInteger core = 0; core < processorCount; core++) {
+            _cpuUsage[@(core)] = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        }
+
+        _systemWrapper = systemWrapper;
+        _processInfoWrapper = processInfoWrapper;
+
+        _memoryFootprint = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        _thermalState = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        _powerLevelState = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+        _memoryPressureState = [NSMutableArray<NSDictionary<NSString *, NSNumber *> *> array];
+
+        _profileStartTime = profileStartTime;
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    [self stop];
+}
+
+#pragma mark - Public
+
+- (void)start
+{
+    [self registerSampler];
+    [self registerStateChangeNotifications];
+    [self registerMemoryPressureWarningHandler];
+}
+
+- (void)stop
+{
+    [_timer invalidate];
+    [_systemWrapper deregisterMemoryPressureNotifications];
+    [_processInfoWrapper stopMonitoring:self];
+}
+
+- (NSMutableDictionary<NSString *, id> *)serialize
+{
+    const auto dict = [NSMutableDictionary<NSString *, id> dictionary];
+
+    if (_memoryFootprint.count > 0) {
```

Comment:
`h`: Are we sure we don't need locking when serializing? Does serialize get called on the same thread as the NSTimer runs, and on the same thread, we receive the notifications?

Reply:
Good call, didn't consider this 👍🏻

Reply:
Notifications are posted on the global dispatch queue, the timer runs on the main runloop, but serialize ultimately comes from `-[SentryTracer finishInternal]` which I'm not sure is always called on the main thread, can you verify that @philipphofmann?

Reply:
Thanks for confirming @philipphofmann, that was my guess.

Reply:
See https://github.com/getsentry/sentry-cocoa/pull/2493/commits/13134a936e618968912586df8bb2b8b84ab7a35d